### PR TITLE
Bind Dynamic Parameters Real-Time

### PR DESCRIPTION
**Pull request by dkattan**
# PR Summary

The goal of this PR is to allow for more flexibility with dynamic parameters by binding RuntimeDefinedParameter objects as they are emitted from the dynamicparam block.

Assuming a helper function exists for creating RuntimeDefinedParameters, you will now be able to define dynamic parameters like this:

```powershell
Function Expand-ComputerName {
    [CmdletBinding()]
    Param()
    DynamicParam {
        Write-Host "Calling New-DynamicParameter for ComputerName"
        New-DynamicParameter -Name ComputerName -Mandatory -Type string
        Write-Host "`$ComputerName: $ComputerName"
        if ($ComputerName -like '*`$AssetTag*') {
            Write-Host "Calling New-DynamicParameter for AssetTag"
            New-DynamicParameter -Name AssetTag -Mandatory -Type string
            Write-Host "Done calling New-DynamicParameter for AssetTag"
        }
    }
    Process {
        $ExecutionContext.InvokeCommand.ExpandString($ComputerName)
    }
}
```
This doesn't break existing functionality as the change is limited to RuntimeDefinedParameter types.

## PR Context

I am building a Web UI for running PowerShell scripts similar to Show-Command.

One of the limitations I encountered when building this UI is that dynamic parameters can only condition off the values of static parameters.

This is because the output of the dynamicParam block is collected and bound at the end.

This change uses the onDataAdded event to process individual RuntimeDefinedParameter objects as they are written. All other object types are accumulated and bound the same as before, preserving existing functionality. 

The second notable change is the creation of variables for these newly bound parameters. 

This is more of a quality of life change to improve the ergonomics of accessing dynamic parameter values.

Historically, you had to retrieve dynamic parameter values from $PSBoundParameters, which feels more like an oversight than an intentional design decision, but I digress.

While the instantiating of variables from string names is not my favorite paradigm as it makes static code analysis more difficult, it is a core PowerShell paradigm.

Invoke-WebRequest uses -SessionVariable and CommonParameters like OutVariable, ErrorVariable and so on create variables based on a specified name. I feel that when a dynamic parameter is found real-time, it makes sense to create a variable for it just like any other bound parameter. Digging into $PSBoundParameters is simply not ergonomic.

The argument could be made that the idea of binding parameters real-time goes against the paradigm of discoverability. But I would argue that when filling out a form visually, it is a lot less daunting to only see parameters that matter for your context, and to that end this facilitates the use of PowerShell.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by dkattan** [comment]
@iSazonov Curious to see your thoughts on this

**Comment by dkattan** [comment]
Any updates from the working group @JamesWTruher  @IISResetMe ?

**Comment by iSazonov** [comment]
> @iSazonov Curious to see your thoughts on this

I tried to come up with a realistic scenario and failed miserably. In my opinion it is almost impossible (and useless) to use it in an interactive scenario. Especially when considering pipeline. As for scripts, it looks like we are moving conditions from outside to inside. We could see the benefit in that, although in my opinion it makes functions and cmdlets too complicated to understand and use. It would be useful to discuss a few real scenarios to see if we can use simpler and easier to understand alternatives.

**Comment by dkattan** [comment]
> > @iSazonov Curious to see your thoughts on this
> 
> I tried to come up with a realistic scenario and failed miserably. In my opinion it is almost impossible (and useless) to use it in an interactive scenario. Especially when considering pipeline. As for scripts, it looks like we are moving conditions from outside to inside. We could see the benefit in that, although in my opinion it makes functions and cmdlets too complicated to understand and use. It would be useful to discuss a few real scenarios to see if we can use simpler and easier to understand alternatives.

Hey @iSazonov thank you for checking it out!

Let me start by showing you what I'm building:
DynamicParameterBinding

The left pane is essentially Show-Command but interactive and on the web.

The fact that I'm using a form necessitates that I collect all parameters from the user ahead of time.

To prevent duplicating the extensive param blocks in the functions responsible for joining Active Directory, joining AzureAD, setting the computer name, and creating users in the respective directories, I instead created re-usable functions like Get-DirectoryTypeParameters that return RuntimeDefinedParameters. I also created a function called Get-CommandParameters that does the same thing in a more generic way.
```powershell
Function Get-CommandParameter
{
  [Cmdletbinding()]
param(
    [Parameter(Mandatory)]
    [ValidateNotNullOrEmpty()]
    [string]
    $Name,
    [Parameter(ValueFromRemainingArguments=$true)]
    [Alias('Args')]
    [AllowNull()]
    [AllowEmptyCollection()]
    $ArgumentList
)
  $CommandMetadata = Get-Command -Name $Name -ErrorAction Stop -ArgumentList $ArgumentList # | %{$_.Parameters.GetEnumerator()} | %{$_}
  $CommonParameterNames = [System.Management.Automation.Internal.CommonParameters].GetProperties().Name
  $CommandMetadata.Parameters.GetEnumerator() | ?{ $_.Key -notin $CommonParameterNames } | %{
    $ParameterName = $_.Key
    [System.Management.Automation.ParameterMetadata]$ParameterMetadata = $_.Value
    New-Object -TypeName System.Management.Automation.RuntimeDefinedParameter($ParameterName, $ParameterMetadata.ParameterType, $ParameterMetadata.Attributes)
  }
}
```

After I created this PR, I discovered I'm not alone in my quest to reuse param block code. RobocopyPS implements New-ProxyFunction to reuse the extensive param block on [Invoke-RoboCopy](https://github.com/sbergwall/RobocopyPS/blob/3c0e11d94b341c4de0ab394ed9170bd2c9e13237/RobocopyPS/internal/New-ProxyFunction.ps1) from functions like [Copy-RoboItem](https://github.com/sbergwall/RobocopyPS/blob/3c0e11d94b341c4de0ab394ed9170bd2c9e13237/RobocopyPS/functions/Copy-RoboItem.ps1) and [Sync-RoboItem](https://github.com/sbergwall/RobocopyPS/blob/3c0e11d94b341c4de0ab394ed9170bd2c9e13237/RobocopyPS/functions/Sync-RoboItem.ps1). 

```powershell
function Copy-RoboItem {
    [CmdletBinding(SupportsShouldProcess = $True)]
    param (
        # Specifies the path to the source directory. Must be a folder.
        [Parameter( Mandatory = $True,
            ValueFromPipelineByPropertyName,
            ValueFromPipeline)]
        [Alias('Path', 'FullPath')]
        [String[]]$Source
    )

    DynamicParam {
        # Get available parameters from Invoke-RoboCopy and ignore parameters that is not for moving items
        New-ProxyFunction -CommandName 'Invoke-RoboCopy' -CommandType 'Function' -ignoredParams "Source", "Purge", "Mirror", "MoveFiles", "MoveFilesAndDirectories"
    }
```

This works because Invoke-RoboCopy doesn't have dynamicparams. Specifically it doesn't have dynamic parameters that are conditionally present based on the value of a paramblock parameter.

My specific use case is to have a single function that sets computer name, joins Active Directory or AzureAD, and creates a profile for the primary person. Naturally each of these tasks are implemented in their own functions complete with parameter validation, dynamic parameters, the works.

Here's a visual:

```mermaid
flowchart TD
subgraph "param()"
$DirectoryType
$UsernameTemplate
$ComputerNameTemplate
end
subgraph "dynamicparams{}"
$DirectoryType --> |AzureAD| $DeviceEnrollmentManagerUsername
$DirectoryType --> |AzureAD| $DeviceEnrollmentManagerPassword
$DirectoryType --> |ActiveDirectory| $DomainSelector
$DomainSelector --> $PrimaryUserOptions
$DirectoryType --> |ValidateSetValues| $UPNSuffix
$DirectoryType --> |ValidateSetValues| $PrimaryUser
$DeviceEnrollmentManagerUsername --> $PrimaryUserOptions
$DeviceEnrollmentManagerPassword --> $PrimaryUserOptions
$PrimaryUserOptions --> |None| End
$PrimaryUserOptions --> |NewUser| NewUser
subgraph NewUser
$UsernameTemplate --> |$FirstName.$LastName| UPNPrefixDefaultValue["Expand-String $UsernameTemplate"] --> |DefaultValue|$UPNPrefix
$UPNPrefix
$FirstName --> UPNPrefixDefaultValue
$LastName --> UPNPrefixDefaultValue
$UPNSuffix
end
end
$PrimaryUserOptions --> |ExistingUser| $PrimaryUser
$PrimaryUser --> End
NewUser --> End
```

Now, imagine I splatted the following Hashtable
```powershell
$ht = @{
   DirectoryType = 'AzureAD'
   DeviceEnrollmentManagerUsername = 'dem@contoso.com'
   DeviceEnrollmentManagerPassword = 'hunter12'
   PrimaryUserOptions = 'NewUser'
   FirstName = 'John'
   LastName = 'Doe'
}
```
Without this change I get ` A parameter cannot be found that matches parameter name 'FirstName'`

This is because `$PrimaryUserOptions` is defined in the dynamicparam block and its value has not yet been bound. 

```powershell
New-DynamicParameter -Name PrimaryUserOptions -ValidateSet 'NewUser','ExistingUser','None' -Mandatory
switch($PrimaryUserOptions)
{
  'NewUser' {
    New-DynamicParameter -Name FirstName -Mandatory -Type string
    New-DynamicParameter -Name LastName -Mandatory -Type string
  }
  'ExistingUser' {
    switch($DirectoryType){
      'ActiveDirectory' {
        $Users = Get-ADUser | select -Expand DisplayName
      }
      'AzureAD' {
        $Users = Get-AzureADUsers | select -Expand DisplayName
      }
    New-DynamicParameter -Name ExistingUser -ValidateSet $Users
  }
}
```

You could argue that I should move `$DirectoryType` and `$PrimaryUserOptions` into the static param block, but what if I decide to combine the functionality of this script with that of another with the use of my Get-CommandParameter function? This again shifts everything back into dynamicparams and I'm back to square one.

In conclusion I'd like this to be considered because
1. It doesn't break anything
2. It eliminates the need to duplicate paramblock dynamicparams blocks when aliasing or combining functions

**Comment by iSazonov** [comment]
@dkattan Thanks for the example. This demonstrates well the wrong approach.
As I said it's moving the logic into parameter binding. This leads to a number of problems. The main one is complexity. I wouldn't want to run into someone else's code that uses this.
Also the UX is very poor. It provokes writing code that may not work well on its own, and in addition it may have a bad effect on Engine. (For example, you already use Get-AzureADUsers for ValidateSet - what if it returns 100000 accounts?)
Plus, pseudo-binding won't work for cascaded dynamic parameters. We would also have to think about PowerShell providers. I don't see how it would work there.
You try to create elegant code, which is always what we all want to achieve, but in reality you create code that collects every conceivable and unthinkable problems.

**Comment by StartAutomating** [comment]
@iSazonov I can see a few other scenarios where this would be useful.

1. Joining dynamic parameters of several commands.
2. More succinctly filtering dynamic parameters
3. (potentially) simplifying conflicts between dynamic parameters

This would simplify some existing code that uses dynamic parameters, such as that in [RoughDraft](https://github.com/StartAutomating/RoughDraft), and the groundwork laid here would open up easier creation of dynamic parameters over time.

As far as the other feedback:

* Attribution and additional metadata are valuable ways to enhance a cmdlet and make it look better in a UI.
* The problems of binding PowerShell to any UI framework are painful and can be made much easier with small changes in the engine, such as this.
* While PowerShell has _lots_ of metadata, that metadata isn't always sufficient to provide a good user experience to a command on its own; it has to be augmented.
* PowerShell providers have their own way of providing dynamic parameters, and it may benefit from a similar change.
* Attributes can be easily stripped or added during the compilation/transpilation of a script (thus it need not clutter everyone's experience of reading the code)
* Last, but not least, it looks like @dkattan put a lot of effort into the UX there, and I don't think it's helpful to insult someone's PowerShell+Web UX in a PR review.

FWIW, I think this would be a pretty useful change.

**Comment by dkattan** [comment]
> This demonstrates well the wrong approach.
> As I said it's moving the logic into parameter binding. This leads to a number of problems. The main one is complexity. I wouldn't want to run into someone else's code that uses this.
I'm trying to move errors from runtime to invocation time. Garbage in, garbage out right? Validation logic _should_ be complex to prevent issues down the line.

> Also the UX is very poor. It provokes writing code that may not work well on its own, and in addition it may have a bad effect on Engine. (For example, you already use Get-AzureADUsers for ValidateSet - what if it returns 100000 accounts?)
There's nothing stopping anyone from creating a ValidateSetAttribute with a string[100000] today. But to be fair, I did encounter this problem and worked around it using caching and pagination on a custom [DropdownAttribute()]

> Plus, pseudo-binding won't work for cascaded dynamic parameters. We would also have to think about PowerShell providers. I don't see how it would work there.
Pseudobinding does work as it calls the cmdletbinder under the hood, I tested it

> You try to create elegant code, which is always what we all want to achieve, but in reality you create code that collects every conceivable and unthinkable problems.

I disagree wholeheartedly. This encourages re-use of validation logic to ensure clean input.

**Comment by TitanDeploy** [comment]
Also an ImmyBot user, this UX is actually pretty good when compared to others I've encountered. Would love to have this functionality available, I can see it coming in handy for more than just the Configure Directory task.

**Comment by bironeaj** [comment]
I think all of the above comments show there is a need for this in PowerShell. As a PowerShell developer and ImmyBot user, I can see this being extremely useful.

@iSazonov, I'm not really sure why the insulting comment about the UI was necessary here. The GIF/screenshot was provided by @dkattan, as you requested a real-world scenario. This PR was opened to discuss a feature being added to PowerShell, not the UI of an unrelated application. This comment is a textbook example of a Microsoft code of conduct violation, which I'll be reporting.

**Comment by ryanleewilliams** [comment]
I can't say I understand all the implications of the change, but it doesn't sound like there are many downsides and it seems to open up a lot of possibilities. I know it's not worth much, but as someone who loves working with PowerShell and also loves working with ImmyBot, I'd like to see this change given further consideration.

**Comment by MWGMorningwood** [comment]
As an ImmyBot user, the UX is extremely easy to grasp once you're in the right mindset.
The team at Immense are automation gods. If they're requesting new features, it's likely for the best.
Their use case of this specific PR is especially relevant to us.

**Comment by iSazonov** [comment]
@dkattan 
> > Plus, pseudo-binding won't work for cascaded dynamic parameters. We would also have to think about PowerShell providers. I don't see how it would work there.
> 
> Pseudobinding does work as it calls the cmdletbinder under the hood, I tested it.

To be clear I don't mean this break pseudobinding, I mean pseudobinding cannot benefit from the enhancement.

> 
> > You try to create elegant code, which is always what we all want to achieve, but in reality you create code that collects every conceivable and unthinkable problems.
> 
> I disagree wholeheartedly. This encourages re-use of validation logic to ensure clean input.

I'm not criticizing your product, much less you as some have written here. I am trying to understand how you benefit from this extension on your example.
Did I understand you correctly that you create a generic cmdlet with a full set of dynamic parameters and then read the meta information and dynamically build the Web interface?

**Comment by JerBar** [comment]
> @dkattan
> 
> > > Plus, pseudo-binding won't work for cascaded dynamic parameters. We would also have to think about PowerShell providers. I don't see how it would work there.
> > 
> > 
> > Pseudobinding does work as it calls the cmdletbinder under the hood, I tested it.
> 
> To be clear I don't mean this break pseudobinding, I mean pseudobinding cannot benefit from the enhancement.
> 
> > > You try to create elegant code, which is always what we all want to achieve, but in reality you create code that collects every conceivable and unthinkable problems.
> > 
> > 
> > I disagree wholeheartedly. This encourages re-use of validation logic to ensure clean input.
> 
> I'm not criticizing your product, much less you as some have written here. I am trying to understand how you benefit from this extension on your example. Did I understand you correctly that you create a generic cmdlet with a full set of dynamic parameters and then read the meta information and dynamically build the Web interface?


I would think that this is more in line with how PowerShell was envisioned to pass metadata back so you have better understanding of the objects you are manipulating between commands, very similar to how the Exchange power shell provider works.

**Comment by IISResetMe** [comment]
The WG-Engine working group recently review this PR. As the current commits effectively cover two functional changes, I'll address them separately:

 - Allow `dynamicparam` to emit runtime parameters in a stream-like fashion
 - Dynamically re-bind parameters to allow for interdependent/cascading runtime parameter generation
 
 ------
 
 ## Streaming output:
 
 Review Verdict: **approved**, with comments
 
We recognize the frustration that the current rigid output requirements for the `dynamicparam` block poses, and welcome this suggested improvement. 

In order to not break existing valid `dynamicparam` blocks, the implementation should observe the following rules:

 - If the blocks outputs a single `RuntimeDefinedParameterDictionary` instance, the existing behavior should persist and the dictionary should be used
 - If the output consists of 1 or more **named** `RuntimeDefinedParameter` instances, construct the dictionary from those

Should you wish to support scenarios deviating from those (eg. allowing mixed-type output and ignoring non-parameters), the functionality must be guarded by an experimental feature flag.

 ------
 
## Dynamic re-bind
 
 Review verdict: **rejected**

While we do recognize that a limited number of use cases might benefit from this ability(and the currently suggested solution is _clever_), we also agree that support for cascading runtime parameter generation deviates too far from the original implementation of dynamic parameters to merit approval as a simple change - `dynamicparams` is already too much of a "blackbox", and we'd like to ensure we don't perpetuate that state of affairs.

Before considering such a drastic change, we would like to see: 

 - A more detailed problem statement (current limitations, core use cases, performance considerations, etc.)
 - Broader input from the community (additional use cases, discussion of additional risk factors)
 - A larger test suite

For these reasons, we recommend **opening a separate issue** to discuss what the implementation might look like and what the broader implications might be

------

@dkattan I appreciate this is not the outcome you were hoping for - please continue sharing your thoughts on the topic as you already have in this discussion. Once we all have a good idea of exactly what limitations the current behavior imposes and what implications the suggested change actually has, we'll happily pick it up for review again.

**Comment by StartAutomating** [comment]
> While we do recognize that a limited number of use cases might benefit from this ability(and the currently suggested solution is _clever_), we also agree that support for cascading runtime parameter generation deviates too far from the original implementation of dynamic parameters to merit approval as a simple change - `dynamicparams` is already too much of a "blackbox", and we'd like to ensure we don't perpetuate that state of affairs.

A few notes here, from a concerned community member:

1. As someone who uses dynamic parameters _a lot_, automatically populating the variable if it does not already exist would be _huge_.  Despite writing a lot of dynamic parameter functionality and knowing "better", I always find myself presuming that the variable will be there, and then copying/pasting the same few lines that do the same thing as this PR (populate the variables in $psBoundParameters).

2. Please don't underestimate the value of fixing this part of the ecosystem.  Dynamic Parameters are incredibly powerful, especially when it command extensibility and reuse.  Being able to "carry" parameters between one command and the next is very useful, and would get a lot easier with this PR (and easier still with a _little_ more sweat)

3. I agree that dynamicparams are too much of a black box.  This is our opportunity to bring sunlight to that black box and make it much easier to work with.  Even if we cannot illuminate the entire black box, clearly documenting how this part works would be a big benefit.  As the saying goes, sunlight is the best disinfectant.

I believe I will try to help improve the testability and transparency in this PR, as I think, done right it would be a "killer feature" for PowerShell Core.

**Comment by SeeminglyScience** [comment]
> 1. As someone who uses dynamic parameters _a lot_, automatically populating the variable if it does not already exist would be _huge_.  Despite writing a lot of dynamic parameter functionality and knowing "better", I always find myself presuming that the variable will be there, and then copying/pasting the same few lines that do the same thing as this PR (populate the variables in $psBoundParameters).

For the record, this was not an aspect of the PR that we discussed. I assumed this part of the external `New-DynamicParameter` function and not actually something in the PR. You're welcome to create an issue for that aspect to be discussed.

> 2. Please don't underestimate the value of fixing this part of the ecosystem.  Dynamic Parameters are incredibly powerful, especially when it command extensibility and reuse.  Being able to "carry" parameters between one command and the next is very useful, and would get a lot easier with this PR (and easier still with a _little_ more sweat)

Command extensibility and reuse is definitely valuable and certainly something worth looking into. I would caution against looking to dynamic parameters as the solution though. Personally, I'd prefer something similar to what we already have in C# (inheritance). 

> 3. I agree that dynamicparams are too much of a black box.  This is our opportunity to bring sunlight to that black box and make it much easier to work with.  Even if we cannot illuminate the entire black box, clearly documenting how this part works would be a big benefit.  As the saying goes, sunlight is the best disinfectant.

I believe Mathias was more referring to the complexity of its implementation in the parameter binder. But you're more than welcome to file an issue in MicrosoftDocs/PowerShell-Docs

**Comment by dkattan** [comment]
> Command extensibility and reuse is definitely valuable and certainly something worth looking into. I would caution against looking to dynamic parameters as the solution though. Personally, I'd prefer something similar to what we already have in C# (inheritance).

What if we did this?

```powershell
Function Get-WhateverFunction : BaseFunction
{

}
```

**Comment by StartAutomating** [comment]
> Yep! That's the idea

FWIW, I also love this idea (and would want to see it in another PR).

I'd like to keep this PR purely dynamic parameters.

FWIW pt 2: when discussing function inheritance in PowerShell, dynamic parameters are a useful building block (they allow for dynamic inheritance of whatever a command's signature is now, as opposed to static inheritance, which will lock to the signature at build time).

> For the record, this was not an aspect of the PR that we discussed. I assumed this part of the external New-DynamicParameter function and not actually something in the PR. You're welcome to create an issue for that aspect to be discussed.

To clarify the record here, Binding a DynamicParameter to a $Variable is a key part of this PR (it's mentioned in the PR summary at the top).

I think that there are other issues within dynamic parameters that should be filed and fixed as part of this PR.  I will file some and mention them on this thread before adding fixes.

The one that's bugging me most is:

```PowerShell
# Creating a dynamic parameter with no attributes fails with the obtuse error:
# Parameter 'Foo' cannot be specified in parameter set '__AllParameterSets'.
$dynamicDictionary.Add(
    "Foo", [Management.Automation.RuntimeDefinedParameter]::new("Foo", [string], $null)
)

```

With a close second being:

```PowerShell
# valid syntax, and adds it to the dictionary, but does not recognize the parameter.

$dynamicDictionary += @{
    "Foo" = [Management.Automation.RuntimeDefinedParameter]::new("Foo", [string], @(
                            $paramAttr = [Parameter]::new()
                            $paramAttr.ParameterSetName = $PSCmdlet.ParameterSetName
                            $paramAttr
                        ))
}
```

Let's put some sunlight on this space and disinfect it!

**Comment by SeeminglyScience** [comment]
> I'd like to keep this PR purely dynamic parameters.

Yep it still is, this PR would definitely not be the place to discuss the implementation details of that proposal. Just a side note in response to one of the stated goals of this PR.

> FWIW pt 2: when discussing function inheritance in PowerShell, dynamic parameters are a useful building block (they allow for dynamic inheritance of whatever a command's signature is now, as opposed to static inheritance, which will lock to the signature at build time).

Both can be true, you can still call into the base command to get its dynamic parameters for instance. The static nature of inheritance is a big part of why I think it's ideal as tooling would be able to pick up on it significantly easier.

> To clarify the record here, Binding a DynamicParameter to a $Variable is a key part of this PR (it's mentioned in the PR summary at the top).

@dkattan correct me if I'm wrong but the main change you wanted was chaining dynamic parameters right? If you'd like to shift this PR to be simply adding variable creation I can add that to the next WG meeting agenda. I can also just create a new issue for it if you'd prefer.

> I think that there are other issues within dynamic parameters that should be filed and fixed as part of this PR. I will file some and mention them on this thread before adding fixes.

Generally it's best to keep PRs as small as possible, it makes it easier to track what the actual changes, keeps review time lower, reduces potential for conflicts with other PRs, and makes it so the WG can clear a single PR rather than "yes, but only if you remove these other changes".

> The one that's bugging me most is:
>
> ... 
>
> With a close second being:
> 
> ...

I know that those changes are related to dynamic parameters, but please make separate issues for discussion. They are out of scope for this PR.

**Comment by StartAutomating** [comment]
> In the mean time I believe @StartAutomating has a transpiler project that can give similar (or exactly?) the same experience as described (disclaimer: I have not used it personally).

@SeeminglyScience thanks for revising a post and mentioning this!

Indeed I do.

[[inherit] transpiler](https://github.com/StartAutomating/PipeScript/blob/main/Transpilers/Inherit.psx.ps1)

Some examples of it "in the wild"

[Inheriting SVG elements](https://github.com/StartAutomating/PSSVG/blob/main/Commands/Shapes/SVG.Kite.ps.ps1)
[Inheriting OBS commands](https://github.com/StartAutomating/obs-powershell/blob/main/Commands/Sources/Set-OBSWindowSource.ps.ps1)

Command inheritance has proven incredibly valuable! This PR would make it even better due to the variable binding.

(hence why I care so much)

As for the other feedback:

> Generally it's best to keep PRs as small as possible, it makes it easier to track what the actual changes, keeps review time lower, reduces potential for conflicts with other PRs, and makes it so the WG can clear a single PR rather than "yes, but only if you remove these other changes".

Heard loud and clear.

I'm primarily working on adding tests, inner docs, and clearing up edge cases around this PR.  The items I mentioned can easily be split off into another PR/feature set.  Will do!

**Comment by dkattan** [comment]
> The WG-Engine working group recently review this PR. As the current commits effectively cover two functional changes, I'll address them separately:
> 
> * Allow `dynamicparam` to emit runtime parameters in a stream-like fashion
> * Dynamically re-bind parameters to allow for interdependent/cascading runtime parameter generation
> 
> ## Streaming output:
> Review Verdict: **approved**, with comments
> 
> We recognize the frustration that the current rigid output requirements for the `dynamicparam` block poses, and welcome this suggested improvement.
> 
> In order to not break existing valid `dynamicparam` blocks, the implementation should observe the following rules:
> 
> * If the blocks outputs a single `RuntimeDefinedParameterDictionary` instance, the existing behavior should persist and the dictionary should be used
> * If the output consists of 1 or more **named** `RuntimeDefinedParameter` instances, construct the dictionary from those
> 
> Should you wish to support scenarios deviating from those (eg. allowing mixed-type output and ignoring non-parameters), the functionality must be guarded by an experimental feature flag.
> 
> ## Dynamic re-bind
> Review verdict: **rejected**
> 
> While we do recognize that a limited number of use cases might benefit from this ability(and the currently suggested solution is _clever_), we also agree that support for cascading runtime parameter generation deviates too far from the original implementation of dynamic parameters to merit approval as a simple change - `dynamicparams` is already too much of a "blackbox", and we'd like to ensure we don't perpetuate that state of affairs.
> 
> Before considering such a drastic change, we would like to see:
> 
> * A more detailed problem statement (current limitations, core use cases, performance considerations, etc.)
> * Broader input from the community (additional use cases, discussion of additional risk factors)
> * A larger test suite
> 
> For these reasons, we recommend **opening a separate issue** to discuss what the implementation might look like and what the broader implications might be
> 
> @dkattan I appreciate this is not the outcome you were hoping for - please continue sharing your thoughts on the topic as you already have in this discussion. Once we all have a good idea of exactly what limitations the current behavior imposes and what implications the suggested change actually has, we'll happily pick it up for review again.

Since the name of this PR was focused on Real-Time binding, I created a separate PR for the approved part which is being able to declare dynamic parameters from a script without encapsulating them in a RuntimeDefinedParameterDictionary.

**Comment by tylernocwing** [comment]
I hope this gets implemented dkattan

**Comment by dlfoster311** [comment]
I would love to see this implemented.
